Source/WebCore/page/FrameView.cpp

```diff
@@ -2457,20 +2457,49 @@ void FrameView::textFragmentIndicatorTimerFired()
     ASSERT(frame().document());
     auto& document = *frame().document();
     
-    if (!m_pendingTextFragmentIndicatorRange)
+    if (!m_pendingTextFragmentIndicatorRange) {
+        cancelScheduledTextFragmentIndicatorTimer();
         return;
+    }
     
-    if (m_pendingTextFragmentIndicatorText != plainText(m_pendingTextFragmentIndicatorRange.value()))
+    if (m_pendingTextFragmentIndicatorText != plainText(m_pendingTextFragmentIndicatorRange.value())) {
+        cancelScheduledTextFragmentIndicatorTimer();
         return;
+    }
     
     auto range = m_pendingTextFragmentIndicatorRange.value();
     TemporarySelectionChange selectionChange(document, { range }, { TemporarySelectionOption::DelegateMainFrameScroll, TemporarySelectionOption::RevealSelectionBounds, TemporarySelectionOption::UserTriggered });
     
     auto textIndicator = TextIndicator::createWithRange(range, { TextIndicatorOption::DoNotClipToVisibleRect }, WebCore::TextIndicatorPresentationTransition::Bounce);
-    if (textIndicator)
-        document.page()->chrome().client().setTextIndicator(textIndicator->data());
+    
+    auto* page = frame().page();
     
     cancelScheduledTextFragmentIndicatorTimer();
+    
+    if (!page)
+        return;
+    
+    if (textIndicator) {
+        constexpr OptionSet<HitTestRequest::Type> hitType { HitTestRequest::Type::ReadOnly, HitTestRequest::Type::Active, HitTestRequest::Type::AllowVisibleChildFrameContentOnly };
+        
+        auto firstCharRange = resolveCharacterRange(range, { 0, 1 });
+        auto firstCharRect = RenderFlexibleBox::absoluteTextRects(firstCharRange);
+    
+        HitTestResult result;
+        result = page->mainFrame().eventHandler().hitTestResultAtPoint(LayoutPoint(firstCharRect.first().center()), hitType);
+        if (!intersects(range, *result.targetNode()))
+            return;
+
+        if (range.endOffset() > 1) {
+            auto lastCharRange = resolveCharacterRange(range, { range.endOffset() - 1, 1 });
```

Comment:
I am pretty sure this is wrong, since endOffset is in the endContainer's space, and resolveCharacterRange's CharacterRange is in the space of all characters in the range.

Reply:
Maybe rniwa or darin know the "right" way to do this!

Reply:
(and by "wrong", I mean you are picking a totally random end character)